fix(FileStatusList): sort copied before renamed

gerhardol opened the pull request:
## Proposed changes

sort copied before renamed as well as before added/removed as copied may have changes worth reviewing.

Order unequal/same/b/a in the same way as they are displayed.

Note: My primary feature is to see Copied (that may have changes) before Renamed (just moved) to simplify reviewing.
As Copied has "changes" I find it natural to see that after Modified, and Renamed should follow Copied.
But Added/Removed could be moved before Copied/Renamed

## Screenshots <!-- Remove this section if PR does not change UI -->

### Before

![image](https://github.com/user-attachments/assets/39434ecf-fc4f-4a74-819a-ecaff2c2e736)

### After

![image](https://github.com/user-attachments/assets/52d48cf8-52e9-4dda-9dc4-b36eed63ff8a)

## Test methodology <!-- How did you ensure quality? -->

Manual

## Merge strategy

I agree that the maintainer squash merge this PR (if the commit message is clear).

----

:black_nib: I contribute this code under [The Developer Certificate of Origin](../blob/master/contributors.txt).

mstv commented on src/app/GitUI/UserControls/FileStatusList.cs:
```diff
@@ -181,36 +181,38 @@ private static ImageListData CreateImageListData()
             };
 
+            // Note: The order of images also affects sorting by file status.
+            // e.g. for for Unequal: Modified, Copied, Renamed etc, then Same: etc
```
```suggestion
            // e.g. for Unequal: Modified, Copied, Renamed, etc, then Same: etc
```

mstv reviewed (approved):
Have not run.

My first quick thoughts:
Copied is kind of an "added" file.
And it is important to me whether it is an exact copy / rename (100% similarity) or not.

So the sorting of the existing file status [icons] makes good sense to me.
(I need to move the Equal filter toolbar button.)

Would it suffice to move the submodule icons up in order to present them more prominently?

gerhardol commented:
> Have not run.
> 
> My first quick thoughts: Copied is kind of an "added" file. And it is important to me whether it is an exact copy / rename (100% similarity) or not.
> 
> So the sorting of the existing file status [icons] makes good sense to me. (I need to move the Equal filter toolbar button.)
> 
> Would it suffice to move the submodule icons up in order to present them more prominently?

Also Copied can be 100% and Modified can be copied. It is not always obvious what Git presents the changes as..,

I interpret above as you support copied before renamed, but also copied/renamed before added/removed?

From a review perspective the logical order would be modified, copied, added, renamed, removed, but copied/removed are related as added/removed (latter could be copied that failed to be recognized as such).

I would prefer to have submodule changes presented as e.g. modified. Previously, they were embedded (just no special watermarked icon), fine for me. Now it seems like they are last. No major issue, not that many changes.

mstv commented:
> I interpret above as you support copied before renamed, but also copied/renamed before added/removed?

It heavily depends on the similarity.
An exact rename is least important.
A rename with modifications could be considered most important.

Added and removed should be adjacent because the Git's rename/copy detection is too limited.

Let's use this PR as it is.

Submodules should be sorted in front of FileStatusModified* in a follow-up (moving the status icons did not have an effect).

gerhardol commented:
Same should be moved last too, also not affected by icon order

gerhardol commented:
> A  rename with modifications could be considered most important.

Normally seen as a copy

mstv commented:
> Same should be moved last too

matches my expectation, but was not important enough to start a discussion

> also not affected by icon order

must not and cannot because of

![image](https://github.com/user-attachments/assets/93349759-4879-44f2-9b8f-315d26e7392d)

and

![image](https://github.com/user-attachments/assets/11e53137-ee60-4945-a539-353242fc291c)

mstv commented:
> > > (moving the status icons for submodules did not have an effect).

This might be caused by asynchronous retrieval of the submodule status. Workaround: Create submodule nodes with a default submodule icon at once, update them asynchronously.

> >
> > also not affected by icon order
> 
> must not and cannot because of

This means changing the icon order cannot affect the sorting of the DiffBranchStatus, which is the primary, independent sort criterion. Sorting the enum values should suffice. Side effects?

gerhardol commented:
> > Same should be moved last too
> 
> matches my expectation, but was not important enough to start a discussion

As I changed the order of icons and mentioned it here, I changed it, updated commit message and description.

For submodules: I have no good repro now, will open an issue if I recreate the problem

mstv reviewed (approved):
👍

---

Regarding submodules:
- in a follow-up
- async status seems to not be the root cause
- works much better if _all_ icons are moved, particularly `SubmodulesManage` and `FolderSubmodule`
- status is `DiffBranchStatus.Unknown` which is sorted behind `SameChange`

gerhardol commented:
> Regarding submodules:
> 
>     * in a follow-up

👍

Plan to merge in some days, just finishing up the discussion here.

>     * async status seems to not be the root cause

That is as before, I believe.
Unequal/B/A/Same is set when the diffs are calculated, so the first key is set anyway.

>     * works much better if _all_ icons are moved, particularly `SubmodulesManage` and `FolderSubmodule`

I believe it is fine having modified/copied/renamed/added/removed/submodule(s), preferable in Unequal to see conflicts easily (but at least now I find it no major issue to see all submodule changes after same).
More icons to tune sorting is not of much help (well maybe the Unequal watermark...)

>     * status is `DiffBranchStatus.Unknown` which is sorted behind `SameChange`

That would mean that `GitItemStatus.DiffStatus` is overwritten, as it is set initially in src\app\GitUI\UserControls\FileStatusDiffCalculator.cs:242
